javascript/ql/src/Security/CWE-830/FunctionalityFromUntrustedSource.ql

```diff
@@ -0,0 +1,157 @@
+/**
+ * @name Inclusion of functionality from untrusted source.
+ * @description Including functionality from an untrusted source may allow
+ *              an attacker to control the functionality and execute arbitrary code.
+ * @kind problem
+ * @problem.severity warning
+ * @security-severity 6.0
+ * @precision high
+ * @id js/functionality-from-untrusted-source
+ * @tags security
+ *       external/cwe/cwe-830
+ */
+
+import javascript
+
+/** A location that adds a reference to an untrusted source. */
+abstract class AddsUntrustedUrl extends Locatable {
+  /** Gets an explanation why this source is untrusted. */
+  abstract string getProblem();
+}
+
+module StaticCreation {
+  /** Holds if `host` is an alias of localhost. */
+  bindingset[host]
+  predicate isLocalhostPrefix(string host) {
+    host.toLowerCase()
+        .regexpMatch([
+            "(?i)localhost(:[0-9]+)?/.*", "127.0.0.1(:[0-9]+)?/.*", "::1/.*", "\\[::1\\]:[0-9]+/.*"
+          ])
+  }
+
+  /** Holds if `url` is a url that is vulnerable to a MITM attack. */
+  bindingset[url]
+  predicate isUntrustedSourceUrl(string url) {
+    exists(string hostPath | hostPath = url.regexpCapture("(?i)http://(.*)", 1) |
+      not isLocalhostPrefix(hostPath)
+    )
+  }
+
+  /** Holds if `url` refers to a CDN that needs an integrity check - even with https. */
+  bindingset[url]
+  predicate isCdnUrlWithCheckingRequired(string url) {
+    // Some CDN URLs are required to have an integrity attribute. We only add CDNs to that list
+    // that recommend integrity-checking.
+    url.regexpMatch("(?i)" +
+        [
+          "^https?://code\\.jquery\\.com/.*\\.js$", "^https?://cdnjs\\.cloudflare\\.com/.*\\.js$",
+          "^https?://cdnjs\\.com/.*\\.js$"
+        ])
```

Comment:
```suggestion
    url.regexpMatch("(?i)^https?://" +
        [
          "code\\.jquery\\.com", "cdnjs\\.cloudflare\\.com",
          "cdnjs\\.com"
        ] + "/.*\\.js$")
```
The common prefix/suffix can be refactored out.  
(The autoformatter will definitely complain about the above suggestion).